### DISCUSSION TITLE
Override <option></option> template generation to add data-attribute

### DISCUSSION THREAD

**Discussion by ghost** [orchidjs/tom-select#444] [Q&A] [answered]
Hi,

Is there a way to override select's options generation? I'm not talking about the visual aspect but really the 
`<option>` tag.

**Motivation**

I have an ajax select where I want to be able to add custom data-attributes into them like so

```
<select multiple="multiple">
  <option data-something="true">Item 1</option>
  <option data-something="false">Item 2</option>
  <option data-something="true">Item 3</option>
</select>
```

and then I want to be able to 

```
dropdown.querySelectorAll("option:checked").forEach((o) => {
   if (o.dataset.something === "true") {
      // do something
   }
});
```

Thank you

**Comment by ghost** [comment, marked as answer]
I found a workaround that works perfectly for me !

```
onChange: function (value) {
      Object.values(this.options).forEach((ajaxOption) => {
        // Could be duplicated because of how ts behave.
        const originalOptions = this.input.querySelectorAll(`option[value="${ajaxOption.value}"]`);
        originalOptions.forEach((originalOption) => {
          Object.keys(ajaxOption).filter((k) => k.startsWith("data-")).forEach((key) => {
            originalOption.dataset[key.replace("data-", "")] = ajaxOption[key];
          });
        });
      });
    }
```

Backend usage

```
render json:

{ items: results.map { |r| { text: r.name, value: r.id, "data-other": r.other? } } }
```

data-other attribute will be appended automatically to <option> tags

**Reply by iamB0rgy** [reply]
This doesn't work for me.

**Reply by ve3** [reply]
This is working!

I think it needs more describe.

1. The code `ajaxOption.value` must match your AJAX value that was set to `<option value="...here...">`. In my case, I returns `building_id` object in JSON and use this in the `option`'s value. If check in my `TomSelect(element, {})` option on second argument, the option `valueField:` was set to `'building_id'`. So, in my code `ajaxOption.value` must be changed to `ajaxOption.building_id`.
2. The code line `Object.keys(ajaxOption).filter((k) => k.startsWith("data-")).forEach((key) => {` means that the AJAX response data has at least a property start with `data-` or it was prepend in AJAX `.then()` process. So, in my case there are `building_id`, `building_name`, `$something_from_tomselect`. I just replace `k.startsWith()` with `!k.startsWith('$')`. And remove `.replace("data-", "")` because it is not need for me.

And done! It's working!

Thanks. ❤️
